### ISSUE TITLE
Linux ARM64 Support

### ISSUE THREAD

**Issue by spham-amzn** [PopcornFX/O3DEPopcornFXPlugin#47]
O3DE has enabled support for Linux ARM64. The current PFX binary that is downloaded for Linux support is x86_64 based, and the binary will cause a build failure on Linux ARM64. In order to support PFX on Linux ARM64, the PFX Gem needs to check the `CMAKE_SYSTEM_PROCESSOR` variable for `aarch64` and handle it appropriately

**Comment by ArthurPKFX** [comment]
Hi,
We'll start looking into this, thanks for bringing it to our attention

**Comment by ArthurPKFX** [comment]
Hi, just a quick update on this. We were able to build our runtime and its test samples for Linux ARM64. We're now waiting on an arm machine we ordered to properly test this and then build & test the O3DE plugin too

**Comment by spham-amzn** [comment]
Thats great! Thanks for the update

**Comment by spham-amzn** [comment]
Is there an update on the Linux ARM64 version? With the Popcorn Gem (@ 1fe9ca34cf930ad4687130dc28df2518c0fad5b1) , getting the following error:
```
In file included from /home/o3de/github/o3de-multiplayersample/build/linux/External/O3DEPopcornFXPlugin-8e237e94/Code/CMakeFiles/PopcornFX.Static.dir/Unity/unity_4_cxx.cxx:3:
In file included from /home/o3de/github/o3de-multiplayersample-assets/Gems/O3DEPopcornFXPlugin/Code/Source/Components/Helpers/PopcornFXHelperLoopEmitterGameComponent.cpp:6:
In file included from /home/o3de/github/o3de-multiplayersample-assets/Gems/O3DEPopcornFXPlugin/Code/Source/PopcornFX_precompiled.h:15:
In file included from /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime/pkapi/include/pk_precompiled_default.h:2:
In file included from /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime/include/popcornfx.h:17:
In file included from /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime/include/popcornfx_base.h:88:
/home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime/include/pk_defs.h:224:11: fatal error: 'pk_arch_arm.h' file not found
#       include "pk_arch_arm.h"

```
The prebuilt binary that it is referencing is `PopcornFX-2.17.3-O3DE-linux`
```
...
/home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime -isystem /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux/PopcornFX/Runtime/include
```

**Comment by spham-amzn** [comment]
@ArthurPKFX Any update on this?

**Comment by ArthurPKFX** [comment]
There was a bit when this was put on the back-burner but there has been great progress and we have an early access ready, we'll send a download link here soon. This is still quite untested, we'd appreciate if you could try using it and tell us if you find issues 🙂

**Comment by ValPKFX** [comment]
I pushed the needed modifications in PAL_linux.cmake on the branch feature/linux-arm64-preview.
This should download the package PopcornFX-2.17.3-O3DE-linux-aarch64.tar.xz for arm64.

**Comment by spham-amzn** [comment]
> I pushed the needed modifications in PAL_linux.cmake on the branch feature/linux-arm64-preview [0181beb](https://github.com/PopcornFX/O3DEPopcornFXPlugin/commit/0181bebf20835101913b6cfd663762f4d8c19b7e). This should download the package PopcornFX-2.17.3-O3DE-linux-aarch64.tar.xz for arm64.

Thanks!

**Comment by spham-amzn** [comment]
I tried the new branch, and I see the arm code, but am getting package hash issues:
```
INFO] root: -- Downloading package into /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux
[INFO] root: -- Downloading package into /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux
[INFO] root: CMake Error at /home/o3de/github/o3de/cmake/3rdPartyPackages.cmake:287 (message):
[INFO] root:   ly_package: - Unable to get package PopcornFX-2.17.3-O3DE-linux from any
[INFO] root:   download server.  Enable LY_PACKAGE_DEBUG to debug.
[INFO] root: 
[INFO] root:   Error from server https://d3t6xeg4fgfoum.cloudfront.net - 22 - "HTTP
[INFO] root:   response code said error"
[INFO] root: 
[INFO] root:   Error from server s3://ly-dev-3p-packages - 22 - The requested URL returned
[INFO] root:   error: 403
[INFO] root: 
[INFO] root:   Error from server file:///home/o3de/github/3p-package-source/packages - 37
[INFO] root:   - "Couldn't read a file:// file"
[INFO] root: 
[INFO] root:   Error from server https://d3t6xeg4fgfoum.cloudfront.net - 22 - "HTTP
[INFO] root:   response code said error"
[INFO] root: 
[INFO] root:   Error from server https://downloads.popcornfx.com/o3de-packages - 1 -
[INFO] root:   Downloaded successfully, but the file hash did not match expected hash!
[INFO] root: Call Stack (most recent call first):
[INFO] root:   /home/o3de/github/o3de/cmake/3rdPartyPackages.cmake:499 (ly_package_internal_download_package)
[INFO] root:   /home/o3de/github/o3de/cmake/3rdPartyPackages.cmake:589 (ly_force_download_package)
[INFO] root:   /home/o3de/github/o3de/cmake/3rdPartyPackages.cmake:691 (ly_enable_package)
[INFO] root:   /home/o3de/github/o3de/cmake/LYWrappers.cmake:583 (ly_download_associated_package)
[INFO] root:   /home/o3de/github/o3de/cmake/LYWrappers.cmake:266 (ly_parse_third_party_dependencies)
[INFO] root:   /home/o3de/github/o3de-multiplayersample-assets/Gems/O3DEPopcornFXPlugin/Code/CMakeLists.txt:40 (ly_add_target)
[INFO] root: 
[INFO] root: 
[INFO] root: -- Configuring incomplete, errors occurred!
[INFO] root: See also "/home/o3de/github/o3de-multiplayersample/build/tools/CMakeFiles/CMakeOutput.log".
[INFO] root: See also "/home/o3de/github/o3de-multiplayersample/build/tools/CMakeFiles/CMakeError.log".

```

**Comment by spham-amzn** [comment]
This is the code I found:
```
set(LY_PACKAGE_SERVER_URLS ${LY_PACKAGE_SERVER_URLS} "https://downloads.popcornfx.com/o3de-packages")
set(package_name PopcornFX-${POPCORNFX_VERSION}-${POPCORNFX_LICENSE}-linux)
if(${CMAKE_SYSTEM_PROCESSOR} STREQUAL "x86_64")
	set(pk_package_hash 0624bf8ff38bce909fbedede6647990d527c8f11ada08000cdd535fc8c2ee121)
	set(pk_package_id 2pEhBPX228JQwwAQ)
elseif(${CMAKE_SYSTEM_PROCESSOR} STREQUAL "aarch64")
	set(pk_package_hash 0017707d58f80cac493048386a6d703ccd498515633430da9c63a4ad5a1ed564)
	set(pk_package_id POPCORNFXPACKAGEIDTOREPLACE_ARM64)
else()
	message(FATAL_ERROR "Unsupported linux architecture ${CMAKE_SYSTEM_PROCESSOR}")
endif()
```

**Comment by ValPKFX** [comment]
Sorry it was downloading the PopcornFX-2.17.3-O3DE-linux.tar.xz package instead of PopcornFX-2.17.3-O3DE-linux-aarch64.tar.xz.
It is fixed with the commit d5f54d0 on the branch feature/linux-arm64-preview.
We don't have a linux arm64 pc right now able to run o3de right now so it's hard to test.

**Comment by spham-amzn** [comment]
Getting further. Now getting:
```
[INFO] root: In file included from /home/o3de/github/o3de-multiplayersample/build/tools/External/O3DEPopcornFXPlugin-8e237e94/Code/CMakeFiles/PopcornFX.Static.dir/Unity/unity_4_cxx.cxx:3:
[INFO] root: In file included from /home/o3de/github/o3de-multiplayersample-assets/Gems/O3DEPopcornFXPlugin/Code/Source/Components/Helpers/PopcornFXHelperLoopEmitterGameComponent.cpp:6:
[INFO] root: In file included from /home/o3de/github/o3de-multiplayersample-assets/Gems/O3DEPopcornFXPlugin/Code/Source/PopcornFX_precompiled.h:15:
[INFO] root: In file included from /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux-aarch64/PopcornFX/Runtime/pkapi/include/pk_precompiled_default.h:9:
[INFO] root: In file included from /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux-aarch64/PopcornFX/Runtime/pk_kernel/include/kr_timers.h:84:
[INFO] root: /home/o3de/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux-aarch64/PopcornFX/Runtime/pk_kernel/include/kr_timers.inl:25:11: fatal error: 'pk_kernel/include/platforms/android/kr_timers.impl.h' file not found
[INFO] root: #       include "pk_kernel/include/platforms/android/kr_timers.impl.h"
[INFO] root:                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[INFO] root: 1 error generated.
[INFO] root: ninja: build stopped: subcommand failed.
```
From this block
```
#if             defined(PK_WINAPI)
#       include "pk_kernel/include/platforms/win32/kr_timers.impl.h"
#elif   (defined(PK_LINUX) || defined(PK_MACOSX)) && (PK_CPU_ARCH == PK_CPU_X86)
#       include "pk_kernel/include/platforms/linux/kr_timers.impl.h"
#elif   defined(PK_ANDROID) || (defined(PK_LINUX) && (PK_CPU_ARCH == PK_CPU_ARM))
#       include "pk_kernel/include/platforms/android/kr_timers.impl.h"
#elif   defined(PK_IOS) || (defined(PK_MACOSX) && (PK_CPU_ARCH == PK_CPU_ARM))
#       include "pk_kernel/include/platforms/apple/kr_timers.impl.h"
#elif   defined(PK_ORBIS) || defined(PK_UNKNOWN2)
#       include "pk_kernel/include/platforms/sce/kr_timers.impl.h"
#elif   defined(PK_NX)
#       include "pk_kernel/include/platforms/nx/kr_timers.impl.h"
#else
#       error   unsupported Timers API
#endif
```

The root cause appears to be that the `./PopcornFX/Runtime/pk_kernel/include/platforms/` does not have an android subfolder

```
~/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux-aarch64/PopcornFX/Runtime/pk_kernel/include/platforms$ ls -la
total 16
drwxrwxr-x 4 o3de o3de 4096 Aug 29 15:12 .
drwxrwxr-x 3 o3de o3de 4096 Aug 30 15:29 ..
drwxrwxr-x 2 o3de o3de 4096 Aug 29 15:12 apple
drwxrwxr-x 2 o3de o3de 4096 Aug 29 15:12 linux
```

**Comment by ValPKFX** [comment]
Indeed there was a small miss again.
We were able to get an arm64 linux to compile o3de and the gem is now compiling correctly with 7a4edfd.
Delete your `~/.o3de/3rdParty/packages/PopcornFX-2.17.3-O3DE-linux-aarch64/` to re-download the new package.
We didn't try to run the editor yet, so there might be runtime issue.

**Comment by spham-amzn** [comment]
Thank you! I will keep trying and testing :-)

**Comment by spham-amzn** [comment]
That fixed it (compile wise)

**Comment by ValPKFX** [comment]
The arm64 package will now be provided by default.
It is added in the 2.18.0 version released today.
I was able to test the simulation and it was running fine.
But I had to do it with rhi=null since my arm64 pc could not create a vulkan context.
However there in no platform specific code for the rendering, so it should be good.
I will close this ticket and don't hesitate to open new tickets if you find issues.
